### PR TITLE
Flipping negative dofs in BuildDofToArrays [bugfix/build-dof-to-arrays]

### PR DESCRIPTION
**Pull request by mlstowell**
Quick PR to address issue #2152.

This seems like something that should have been caught with a unit test. I'm not sure of an appropriate use case so if anyone has suggestions please comment or add a suitable test.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2153](https://github.com/mfem/mfem/pull/2153) | @mlstowell | @tzanio | @tzanio + @pazner | 04/04/21 | 04/07/21 | 04/10/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by woutertonnon — APPROVED** [review]

**Comment by woutertonnon** [comment]
I find it difficult to come up with a unit test that covers all the functionality of this method. However, an idea for a unit test would be the following. We consider a simple mesh in 2D with edges on the boundary that are negatively oriented  with respect to the unique element they belong to. (I need to double-check this, but probably star.mesh satisfies this property). Next, we choose a finite-element space (e.g. Nedelec, Raviart-Thomas) and use `FiniteElementSpace::BuildDofsToArrays()` to build the necessary array. We then evaluate this array for every dof index using the method `int FiniteElementSpace::GetElementForDof(int dof_ind)`. If everything is working correctly, we expect the output to always be non-negative. In the case where negative dofs are not processed appropriately, the corresponding element index will be written to an arbitrary place in memory. This leaves the element in the array corresponding to the dof at the default value, i.e. `-1`, and we would catch the error. If you think this test is suitable, I can write the code tonight.

**Comment by mlstowell** [comment]
Hi, @woutertonnon,

Unfortunately, I don't think that scheme will succeed. Negatively oriented boundary elements are caught, and "fixed", by during the mesh construction process.

Perhaps we just need a simple unit test which calls `GetElementForDof` and `GetLocalDofForDof` for every degree of freedom on a small mesh and confirms the results. Or perhaps not so simple. We would need to check various basis function types on various element types both in serial and parallel. I think I have a test with a good collection of meshes but it's not in master yet.  I'll try to put something together.

Thanks for the feedback!
Best wishes,
Mark

File: fem/fespace.cpp [hunk 1]
```diff
@@ -357,10 +357,11 @@ void FiniteElementSpace::BuildDofToArrays()
       const int n = elem_dof -> RowSize(i);
       for (int j = 0; j < n; j++)
       {
-         if (dof_elem_array[dofs[j]] < 0)
+         int dof = (dofs[j] >= 0) ? dofs[j] : - dofs[j] - 1;
```
Comment (pazner) [inline fem/fespace.cpp:360]:
```suggestion
         int dof = DecodeDof(dofs[j]);
```

A bit more self-documenting this way

**Review by pazner — APPROVED** [review]

**Comment by tzanio** [comment]
LGTM

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**